src/Microsoft.ServiceFabric.Actors.KVSToRCMigration/DataValidationWorker.cs

```diff
@@ -0,0 +1,201 @@
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Fabric;
+    using System.Linq;
+    using System.Runtime.Serialization;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.ServiceFabric.Actors.KVSToRCMigration.Models;
+    using Microsoft.ServiceFabric.Actors.Runtime;
+    using Microsoft.ServiceFabric.Data;
+    using Microsoft.ServiceFabric.Services.Communication.Client;
+    using static Microsoft.ServiceFabric.Actors.KVSToRCMigration.MigrationConstants;
+    using static Microsoft.ServiceFabric.Actors.KVSToRCMigration.PhaseInput;
+    using static Microsoft.ServiceFabric.Actors.KVSToRCMigration.PhaseResult;
+
+    internal class DataValidationWorker : WorkerBase
+    {
+        private static readonly string TraceType = typeof(DataValidationWorker).Name;
+        private StatefulServiceInitializationParameters initParams;
+        private ServicePartitionClient<HttpCommunicationClient> servicePartitionClient;
+        private MigrationSettings migrationSettings;
+
+        public DataValidationWorker(
+            KVStoRCMigrationActorStateProvider stateProvider,
+            ActorTypeInformation actorTypeInfo,
+            ServicePartitionClient<HttpCommunicationClient> servicePartitionClient,
+            MigrationSettings migrationSettings,
+            WorkerInput workerInput,
+            string traceId)
+            : base(stateProvider, workerInput, traceId)
+        {
+            this.initParams = this.StateProvider.GetInitParams();
+            this.servicePartitionClient = servicePartitionClient;
+            this.migrationSettings = migrationSettings;
+        }
+
+        public override async Task<WorkerResult> StartWorkAsync(CancellationToken cancellationToken)
+        {
+            ActorTrace.Source.WriteInfoWithId(
+                        TraceType,
+                        this.TraceId,
+                        $"Starting or resuming data validation worker\n Input: {this.Input.ToString()}");
+
+            try
+            {
+                var startSN = this.Input.StartSeqNum;
+                var endSN = this.Input.EndSeqNum;
+
+                var migratedKeys = new List<string>();
+                using (var tx = this.StateProvider.GetStateManager().CreateTransaction())
+                {
+                    var savedMigratedKeys = await this.MetadataDict.TryGetValueAsync(tx, MigrationConstants.MigrationKeysMigrated);
+                    if (savedMigratedKeys.HasValue)
+                    {
+                        migratedKeys = savedMigratedKeys.Value.Split(DefaultDelimiter).ToList();
+                    }
+                }
+
+                var noOfKeysToValidate = Math.Round(migratedKeys.Count * (this.migrationSettings.PercentageOfMigratedDataToValidate / 100), 0);
+                var rand = new Random();
+                var validatedKeysIndices = new List<int>();
+
+                for (long index = startSN; index <= endSN; index++)
```

Comment:
We don't need to loop over all the SNs since we are validating only a subset. We can instead loop over no of keys validated. this would also avoid the check needed in line 90